File: aoc2/main.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

char * trim(char * s) {
    int l = strlen(s);
    while(isspace(s[l - 1])) --l;
    while(* s && isspace(* s)) ++s, --l;
    return strndup(s, l);
}

int main() {
  FILE *file;
  char line[100];
  char *rule, *passwd, letter, *occ_range;
  int valid1 = 0, valid2 = 0, max_occ, min_occ, occ = 0;

  file = fopen("input.txt", "r");

  while(fgets(line, 100, file)) {
    rule = strtok(line, ":");
    passwd = trim(strtok(NULL, ":"));
    letter = rule[strlen(rule) - 1];
    occ_range = strtok(rule, " ");

    min_occ = atoi(strtok(occ_range, "-"));
    max_occ = atoi(strtok(NULL, "-"));
    for(int i = 0; i < strlen(passwd); i++) {
      if(passwd[i] == letter) {
        occ += 1;
        if(occ > max_occ) break;
      }
    }

    if (occ >= min_occ && occ <= max_occ) {
      valid1 += 1;
    }

    if (passwd[min_occ - 1] == letter ^ passwd[max_occ - 1] == letter) {
      valid2 += 1;
    }

    occ = 0;
  }
  fclose(file);
  printf("part1: %d passwords are valid!\n", valid1);
  printf("part2: %d passwords are valid!\n", valid2);
  return 0;
}

File: aoc2/main.py
def readFile(filename):
    with open(filename) as file:
        lines = [line.strip() for line in file]
    file.close()

    return lines


if __name__ == '__main__':
    fileData = readFile('input.txt')

    valid1 = 0
    valid2 = 0

    for line in fileData:
        rule = line.split(':')[0]
        passwd = line.split(':')[1].strip()
        letter = rule[-1]
        occ_range = rule.split(' ')[0]
        low, high = int(occ_range.split('-')[0]), int(occ_range.split('-')[1])

        if low <= passwd.count(letter) <= high:
            valid1 += 1

        if (passwd[low - 1] == letter) ^ (passwd[high - 1] == letter):
            valid2 += 1

    print(f'part1: {valid1} passwords are valid')
    print(f'part2: {valid2} passwords are valid')
